### PR TITLE
Horizontal scroll animations improvements

### LINKED ISSUE #1136: Kinetic animation and reset time scale conflicts
**Lightweight Charts Version:** 3.8

**Steps/code to reproduce:**

* start kinetic animation
* reset time scale via API

**Actual behavior:**

Glitches

**Expected behavior:**

No glitches

**Screenshots:**

https://user-images.githubusercontent.com/1572087/182591383-055452f1-7995-4454-9a63-956a55e51221.mov

### PR DESCRIPTION
**Pull request by kirchet**
**Type of PR:** bugfix

**PR checklist:**

- [x] Addresses an existing issue: fixes #1136
- [ ] Includes tests
- [ ] Documentation update

**Overview of change:**

scroll_animation

### CODE REVIEW

File: src/model/time-scale.ts [hunk 1]
```diff
@@ -668,21 +690,49 @@ export class TimeScale {
...
+				return;
+			}
+
 			const animationProgress = (performance.now() - animationStart) / animationDuration;
 			const finishAnimation = animationProgress >= 1;
 			const rightOffset = finishAnimation ? offset : source + (offset - source) * animationProgress;
 			this.setRightOffset(rightOffset);
 			if (!finishAnimation) {
-				setTimeout(animationFn, 20);
+				requestAnimationFrame(animationFn);
```
Comment (Nipheris) [inline src/model/time-scale.ts:710]: `requestAnimationFrame` in a `TimeScale`.. Sounds more like a job for `PaneWidget`. And, btw, we already have animation logic there, why not make it generic for both types of animation?

I suggest you to move both `KineticAnimation` and this uniform animation objects back to PaneWidget, and use some generic interface for animations like:
- finished(time)
- terminated
- getPosition(time).

Then use `model.scrollTimeTo` or `timeScale.setRightOffset` (not sure what fits better) to update the model according to current animation, like we already do in `PaneWidget`: `model.scrollTimeTo(scrollXAnimation.getPosition(now));`

File: src/model/time-scale.ts [hunk 2]
```diff
@@ -668,21 +690,49 @@ export class TimeScale {
...
 		const animationStart = performance.now();
+
+		let terminated = false;
+
 		const animationFn = () => {
+			if (terminated) {
+				return;
+			}
+
 			const animationProgress = (performance.now() - animationStart) / animationDuration;
```
Comment (Nipheris) [inline src/model/time-scale.ts:705]: A sidenote: it is better to use the `time` argument passed to `animationFn` by `requestAnimationFrame` to get the "current" time. Nevertheless, `performance.now()` should still be used to capture "start" time.

**Review by Nipheris — CHANGES_REQUESTED** [review]

File: src/gui/pane-widget.ts [hunk 3]
```diff
@@ -678,39 +669,31 @@ export class PaneWidget implements IDestroyable, MouseEventHandlers {
...
 		}
 
-		const model = this._model();
-		const timeScale = model.timeScale();
-
-		const scrollXAnimation = this._scrollXAnimation;
-
 		const animationFn = () => {
 			if ((scrollXAnimation.terminated())) {
 				// animation terminated, see _terminateKineticAnimation
```
[the remark this answers is not in the document]
Comment (kirchet) [reply src/gui/pane-widget.ts]: Fixed

File: src/model/time-scale.ts [hunk 4]
```diff
@@ -668,21 +690,49 @@ export class TimeScale {
...
 			const rightOffset = finishAnimation ? offset : source + (offset - source) * animationProgress;
 			this.setRightOffset(rightOffset);
 			if (!finishAnimation) {
-				setTimeout(animationFn, 20);
+				requestAnimationFrame(animationFn);
 			}
 		};
 
+		this._scrollToPositionAnimation = {
+			terminate: () => {
```
[the remark this answers is not in the document]
Comment (kirchet) [reply src/model/time-scale.ts]: https://github.com/tradingview/lightweight-charts/pull/1140/commits/33e10e99d5dbb791b1cf13f4b7891e244d071127

File: src/model/time-scale.ts (hunk 1 above)
Comment (kirchet) [reply src/model/time-scale.ts]: https://github.com/tradingview/lightweight-charts/pull/1140/commits/33e10e99d5dbb791b1cf13f4b7891e244d071127

File: src/model/time-scale.ts (hunk 2 above)
Comment (kirchet) [reply src/model/time-scale.ts]: Fixed

**Review by Nipheris — APPROVED** [review]

**Pull request merged**